packages/eslint-plugin/lib/rules/unified-signatures.js

```diff
@@ -0,0 +1,616 @@
+
+/** @type {import("eslint").Rule.RuleModule} */
+module.exports = {
+  meta: {
+    docs: {
+      description:
+        'Warns for any two overloads that could be unified into one by using a union or an optional/rest parameter.',
+      category: 'TypeScript-specific',
+      recommended: false,
+      extraDescription: [util.tslintRule('unified-signatures')],
+      url: util.metaDocsUrl(' unified-signature')
+    },
+    type: 'suggestion'
+  },
+
+  create: function(context) {
+    const sourceCode = context.getSourceCode();
+
+    /**
+     * @typedef Failure
+     * @type {object}
+     * @property {Unify} unify
+     * @property {boolean} only2
+     */
+    /**
+     * @typedef Unify
+     * @type {{kind: 'single-parameter-difference', p0: any, p1: any} | {kind: 'extra-parameter', extraParameter: any, otherSignature: any}}
+     */
+    /**
+     * Given a node, if it could potentially be an overload, return its signature and key.
+     * All signatures which are overloads should have equal keys.
+     * @template T
+     * @callback GetOverload
+     * @param {T} node
+     * @returns {{signature: any, key: string} | undefined}
+     */
+    /**
+     * Returns true if typeName is the name of an *outer* type parameter.
+     * In: `interface I<T> { m<U>(x: U): T }`, only `T` is an outer type parameter.
+     * @callback IsTypeParameter
+     * @param {string} typeName
+     * @returns {boolean}
+     */
+    /**
+     * @template T
+     * @template Out
+     * @callback ForEachPairCallback
+     * @param {T} a
+     * @param {T} b
+     * @returns {Out | undefined}
+     */
+
+    //----------------------------------------------------------------------
+    // Helpers
+    //----------------------------------------------------------------------
+
+    /**
+     * @param {number} [otherLine]
+     */
+    function FAILURE_STRING_OMITTING_SINGLE_PARAMETER(otherLine) {
+      return `${FAILURE_STRING_START(otherLine)} with an optional parameter.`;
+    }
+
+    /**
+     * @param {number} [otherLine]
+     */
+    function FAILURE_STRING_OMITTING_REST_PARAMETER(otherLine) {
+      return `${FAILURE_STRING_START(otherLine)} with a rest parameter.`;
+    }
+
+    /**
+     * @param {number | undefined} otherLine
+     * @param {string} type1
+     * @param {string} type2
+     */
+    function FAILURE_STRING_SINGLE_PARAMETER_DIFFERENCE(
+      otherLine,
+      type1,
+      type2
+    ) {
+      return `${FAILURE_STRING_START(
+        otherLine
+      )} taking \`${type1} | ${type2}\`.`;
+    }
+
+    /**
+     * @param {number} [otherLine]
+     * @returns {string}
+     */
+    function FAILURE_STRING_START(otherLine) {
+      // For only 2 overloads we don't need to specify which is the other one.
+      const overloads =
+        otherLine === undefined
+          ? 'These overloads'
+          : `This overload and the one on line ${otherLine}`;
+      return `${overloads} can be combined into one signature`;
```

Comment:
messages should be migrated to `messageId`, you can find usages in other rules or in documentation:

https://eslint.org/docs/developer-guide/working-with-rules#messageids

Reply:
Done.